### PR TITLE
IGNITE-12891 Add userAttributes map to all GridClient messages

### PR DESCRIPTION
**Pull request by oleg-ostanin**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AdditionalSecurityCheckTest.java [hunk 1]
```diff
@@ -250,6 +252,29 @@ public void testClientInfoIgniteClientFail() throws Exception {
...
+        assertEquals(2, ignite.cluster().topologyVersion());
+
+        ignite.cluster().state(ClusterState.ACTIVE);
+
+        try (GridClient client = GridClientFactory.start(getGridClientConfiguration())) {
+            assertTrue(client.connected());
+
+            GridClientClusterState state = client.state();
+
+            // Close a coordinator to force the client to send a CLUSTER_CURRENT_STATE message to the other node.
```
Comment (andrey-kuznetsov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AdditionalSecurityCheckTest.java:271]: ... "in the next statement", I'd like to add. It's not so obvious from the first glance, that state.state() causes some communication with cluster.

Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AdditionalSecurityCheckTest.java]: added this to the comment as you suggested

File: modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java [hunk 2]
```diff
@@ -638,6 +644,29 @@ private void removePending(long reqId) {
...
+
+        try {
+            sndFut.get();
+        }
+        catch (Exception e) {
+            throw new GridClientConnectionResetException("Failed to send message over connection " +
+                "(will try to reconnect): " + serverAddress(), e);
+        }
+
+        return fut;
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java:667]: Duplication of `makeRequest`, please reuse `makeRequest` instead.

File: modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java [hunk 3]
```diff
@@ -646,10 +675,16 @@ private void removePending(long reqId) {
     private GridClientAuthenticationRequest buildAuthRequest() {
         GridClientAuthenticationRequest req = new GridClientAuthenticationRequest();
 
+        long reqId = reqIdCntr.getAndIncrement();
+
+        req.requestId(reqId);
+
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java:681]: It's redundant if `makeRequest` is used to send request.

File: modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java [hunk 4]
```diff
@@ -264,6 +264,12 @@ else if (marsh instanceof GridClientJdkMarshaller)
 
             ses.addMeta(SES_META_CONN, this);
 
+            if (cred != null || userAttrs != null) {
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java:267]: `userAttrs != null` -> `!F.isEmpty(userAttrs)`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java [hunk 5]
```diff
@@ -245,19 +245,19 @@
...
-                ses = session(req);
-            }
-            catch (IgniteAuthenticationException e) {
-                return new GridFinishedFuture<>(new GridRestResponse(STATUS_AUTH_FAILED, e.getMessage()));
-            }
-            catch (IgniteCheckedException e) {
-                return new GridFinishedFuture<>(new GridRestResponse(STATUS_FAILED, e.getMessage()));
-            }
+        try {
+            ses = session(req);
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java:251]: Using such an approach you will lose user attributes when session will expire.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java [hunk 6]
```diff
@@ -50,13 +51,23 @@ public void credentials(Object cred) {
     @Override public void writeExternal(ObjectOutput out) throws IOException {
         super.writeExternal(out);
 
+        U.writeString(out, login());
+        U.writeString(out, password());
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:55]: Login and password already included into `cred` object, we don't need to pass it twice.
Also login and password fields are useless and should be removed at all from `GridClientAbstractMessage`
`userAttrs` field should be moved to `GridClientAuthenticationRequest`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java [hunk 7]
```diff
@@ -50,13 +51,23 @@ public void credentials(Object cred) {
...
+
+        U.writeMap(out, userAttrs);
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:57]: Should be serialized after `cred` (perhaps, in this case, compatibility will be supported)

Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:57]: Should be serialized after `cred` (perhaps, in this case, compatibility will be supported)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.security.GridSecurityProcessor;
+import org.apache.ignite.plugin.security.SecurityPermissionSet;
+
+/**
+ *
+ */
+public class TestAttributeSecurityPluginProvider extends TestAdditionalSecurityPluginProvider {
+    /** Authentication handler. */
+    private Consumer<Map<String, Object>> hndlr;
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java:33]: Let's intercept not only user attributes, but whole `authCtx`, in this case, we can reuse this plugin provider in other tests in the future.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+            // Close a coordinator to force the client to send a CLUSTER_CURRENT_STATE message to the other node
+            // in the state.state() statement.
+            ignite.close();
+
+            userAttrs = null;
+
+            state.state();
+
+            for (int i = 0; i < TEST_KEY_CNT; i++)
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:86]: It's enough to check only one custom attribute (or check the whole map of attributes)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    @Test
+    public void testClientInfoGridClientNotFail() throws Exception {
+        Ignite ignite = startGrids(2);
+
+        assertEquals(2, ignite.cluster().topologyVersion());
+
+        ignite.cluster().state(ClusterState.ACTIVE);
+
+        try (GridClient client = GridClientFactory.start(getGridClientConfiguration())) {
+            assertTrue(client.connected());
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:74]: You should also check attributes after the first connection.

Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:74]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java (hunk 8 above)
Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java:33]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java (hunk 9 above)
Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:86]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java (hunk 7 above)
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:57]: fixed

Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:57]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java (hunk 6 above)
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:55]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java (hunk 4 above)
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java:267]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java (hunk 3 above)
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java:681]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java (hunk 2 above)
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java:667]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java (hunk 5 above)
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java]: Moved saving attributes to GridNioSession.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+ */
+@RunWith(JUnit4.class)
+public class AttributeSecurityCheckTest extends CommonSecurityCheckTest {
+    /** */
+    private static Map<String, Object> userAttrs;
+
+    /** {@inheritDoc} */
+    @Override protected PluginProvider<?> getPluginProvider(String name){
+        return new TestAttributeSecurityPluginProvider(name, null, ALLOW_ALL,
+            globalAuth, true, new Consumer<AuthenticationContext>() {
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:49]: Let's simplify handler with lambda.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.function.Consumer;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.security.GridSecurityProcessor;
+import org.apache.ignite.plugin.security.AuthenticationContext;
+import org.apache.ignite.plugin.security.SecurityPermissionSet;
+
+/**
+ *
+ */
+public class TestAttributeSecurityPluginProvider extends TestAdditionalSecurityPluginProvider {
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java:31]: Let's rename it, it's not related only to attributes now.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityProcessor.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.function.Consumer;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.security.SecurityContext;
+import org.apache.ignite.plugin.security.AuthenticationContext;
+
+/**
+ * Security processor for test AuthenticationContext with user attributes.
+ */
+public class TestAttributeSecurityProcessor extends TestAdditionalSecurityProcessor {
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityProcessor.java:30]: Class hasn't used outside of `TestAttributeSecurityPluginProvider`, can be moved to TestAttributeSecurityPluginProvider as inner static class.
Also, should be renamed.

Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityProcessor.java:30]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java (hunk 12 above)
Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestAttributeSecurityPluginProvider.java:31]: renamed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java (hunk 11 above)
Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:49]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java [hunk 14]
```diff
@@ -30,34 +33,55 @@
...
     }
 
     /** {@inheritDoc} */
     @Override public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
         super.readExternal(in);
 
-        cred = in.readObject();
+        cred = (SecurityCredentials) in.readObject();
```
[the remark this answers is not in the document]
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:82]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+    /** {@inheritDoc} */
+    @Override protected GridClientConfiguration getGridClientConfiguration() {
+        return super.getGridClientConfiguration()
+            .setUserAttributes(userAttributes());
+    }
+
+    /** */
+    @Test
+    public void testClientInfoGridClientNotFail() throws Exception {
```
[the remark this answers is not in the document]
Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:58]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+import org.apache.ignite.internal.processors.security.impl.TestAuthenticationContextSecurityPluginProvider;
+import org.apache.ignite.plugin.PluginProvider;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import static org.apache.ignite.plugin.security.SecurityPermissionSetBuilder.ALLOW_ALL;
+
+/**
+ * Security tests for thin client.
```
[the remark this answers is not in the document]
Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:37]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java [hunk 17]
```diff
@@ -264,6 +265,12 @@ else if (marsh instanceof GridClientJdkMarshaller)
 
             ses.addMeta(SES_META_CONN, this);
 
+            if (cred != null || F.isEmpty(userAttrs)) {
```
[the remark this answers is not in the document]
Comment (oleg-ostanin) [reply modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java:268]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 18]
```diff
@@ -34,7 +34,7 @@
 import static org.apache.ignite.plugin.security.SecurityPermissionSetBuilder.ALLOW_ALL;
 
 /**
- * Security tests for thin client.
+ * Checks user attributes presents in CridClient messages.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:37]: CridClient -> GridClient

Comment (oleg-ostanin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/CommonSecurityCheckTest.java [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+
+        cfg.setPluginProviders(getPluginProvider(name));
+
+        SslContextFactory sslFactory = (SslContextFactory) GridTestUtils.sslFactory();
+
+        cfg.setSslContextFactory(sslFactory);
+        cfg.setConnectorConfiguration(new ConnectorConfiguration()
+            .setSslEnabled(true)
+            .setSslClientAuth(true)
+            .setSslClientAuth(true)
```
Comment (zstan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/CommonSecurityCheckTest.java:99]: code duplication here.

**Pull request closed without merging**